src/main/tetris_board.cpp

```diff
@@ -190,3 +247,115 @@ bool TetrisBoard::check_corners( TetrisPiece& piece, int direction_unit )
   }
   return can_shift;
 }
+
+vector<int> TetrisBoard::determine_full_rows()
+{
+  vector<int> full_rows;
+  for( size_t row = 0; row < occupied_spaces.size(); row++ )
+  {
+    bool full_row = true;
+    for( bool square_occupied : occupied_spaces.at( row ) )
+    {
+      if( !square_occupied )
+      {
+        full_row = false;
+      }
+    }
+    
+    if( full_row )
+    {
+      full_rows.push_back( row );
+    }
+  }
+  return full_rows;
+}
+
+void TetrisBoard::eliminate_full_rows( vector<int> full_rows )
+{
+  vector<unique_ptr<RenderComponent>> new_components;
+  for( auto& component : components )
+  {
+    try
+    {
+      Block& block = dynamic_cast<Block&>( *component );
+      vector<int>::iterator iterator = std::find( 
+        full_rows.begin(), 
+        full_rows.end(), 
+        block.get_row() );
+
+      if( iterator == full_rows.end() )
+      {
+        new_components.push_back( move( component ) );
```

Comment:
This could potentially be considered concurrent modification of the components list. However, in this case, the size of the list itself is not changed, and each component that is moved out of the unique_ptr will not be referenced again. I can get away with it, so I will.